internal/generator/generator_util_test.go

```diff
@@ -0,0 +1,166 @@
+}
+
+func getTestSrcCodePath(relativePath string) (string, error) {
+	root, err := getProjectRootAbsPath()
+	if err != nil {
+		return "", err
+	}
+	return path.Join(root, "internal", "generator", "testdata", "source_codes", relativePath), nil
+}
+
+func getExpectedFilePath(relativePath string) (string, error) {
+	root, err := getProjectRootAbsPath()
+	if err != nil {
+		return "", err
+	}
+	relativePath = strings.TrimSuffix(relativePath, ".c")
+	relativePath = strings.TrimSuffix(relativePath, ".cpp")
+	return path.Join(root, "internal", "generator", "testdata", "expected", relativePath), nil
+}
+
+//nolint:gochecknoglobals
+var (
+	update = flag.Bool("update", false,
+		`update the expected output of these tests, 
+only use when the expected output is outdated and you are sure your output is correct`)
+)
+
+func TestMain(m *testing.M) {
+	flag.Parse()
+	os.Exit(m.Run())
+}
+
+//nolint:funlen,gocognit
+func TestGenSupFromSrcCode(t *testing.T) {
+	t.Parallel()
+	tests := []struct {
+		name         string
+		relativePath string
+		wantErr      bool
+	}{
+		{
+			name:         "normalDirectiveDefinition_pass",
+			relativePath: "normalDefinition",
+			wantErr:      false,
+		},
+		{
+			name:         "unknownBitmask_fail",
+			relativePath: "unknownBitmask",
+			wantErr:      true,
+		},
+		{
+			name:         "noDirectivesDefinition_fail",
+			relativePath: "noDirectives",
+			wantErr:      true,
+		},
+		// If one directive was defined in several files, we should keep all
+		// of the bitmask definitions
+		{
+			name:         "directiveRepeatDefine_pass",
+			relativePath: "repeatDefine",
+		},
+		// If there are comments in definition, we should delete them
+		{
+			name:         "commentsInDefinition_pass",
+			relativePath: "commentsInDefinition",
+		},
+		// If there are comments in definition, we should delete them
+		{
+			name:         "genFromSingleFile_pass",
+			relativePath: "single_file.c",
+		},
+	}
+	for _, tc := range tests {
+		tc := tc
+		t.Run(tc.name, func(t *testing.T) {
+			t.Parallel()
+			var err error
+			codePath, err := getTestSrcCodePath(tc.relativePath)
+			if err != nil {
+				t.Fatal(err)
+			}
+
+			var buf bytes.Buffer
+
+			err = genFromSrcCode(codePath, "directives", "Match", &buf)
+
+			if !tc.wantErr && err != nil {
+				t.Fatal(err)
+			}
+
+			if tc.wantErr && err == nil {
+				t.Fatal("expected error, got nil")
+			}
+
+			// If the testcase wants an error and there is an error, skip the output file validation,
+			// since there may not be an output file while error occurs in generation.
+			if err != nil {
+				return
+			}
+
+			expectedFilePth, err := getExpectedFilePath(tc.relativePath)
+			if err != nil {
+				t.Fatal(err)
+			}
+
+			expectedFile, err := os.OpenFile(expectedFilePth, os.O_RDWR|os.O_CREATE, 0644)
+			if err != nil {
+				t.Fatal(err)
+			}
+
+			if *update {
+				_, err = expectedFile.WriteString(buf.String())
+				if err != nil {
+					t.Fatal(err)
+				}
+				expectedFile.Close()
```

Comment:
You should not ignore the error returned by `Close` when you are writing the file. This error may indicate that the write has failed. Ignoring the error when you are only reading is generally OK.

Reply:
There's another potential bug here where `expectedFile` is only closed if `*update == true` and `WriteString` did not return an error. While I don't think it will ever cause an issue and everything will be cleaned up when the test run exits, it's good practice to make sure you aren't leaking file descriptors. 

```go
expectedFile, err := os.OpenFile(expectedFilePth, os.O_RDWR|os.O_CREATE, 0644)
if err != nil {
	t.Fatal(err)
}

defer func() {
	if err := expectedFile.Close(); err != nil {
		t.Fatal(err)
	}
}()

if *update {
	_, err = expectedFile.Write(buf.Bytes()).
	if err != nil {
		t.Fatal(err)
	}
	return
}
```

Reply:
> You should not ignore the error returned by Close when you are writing the file.

Oh I see. Thanks for pointing it out!

> While I don't think it will ever cause an issue and everything will be cleaned up when the test run exits, it's good practice to make sure you aren't leaking file descriptors.

Sure! Committed